### ISSUE TITLE
How far did you get with unpacking uied.templates file for etw? 

### ISSUE THREAD

**Issue by szmania** [taw/etwng#6]
How far did you get with unpacking uied.templates file for etw? 
Looks like the template luac (ie: template.row_template_region.luac) files depend on the ui files inside uied.templates file, to generate ui objects from templates.

**Comment by taw** [comment]
You mean for .luac files some other kind of files?
.luac decompiler is a jerryrigged mess.

**Comment by szmania** [comment]
Thanks for the reply. No mate, I can decompile the luac to lua and recompile to luac no issue. But such lua files as template.row_template_region.lua reference ui keys that I can't find in any of the ui -> xml files. So I'm assuming that those ui keys (or xml objects, whatever they are called) live in uied.templates file in ui/templates/uied.templates. Also, NTW has its template UI failes as regular ui files, which can be converted to xml. But NTW doesn't have uied.templates file. ETW does, which I'm assuming all the ui template files live in.  

Does your  etwng/templates/templates_unpack unpack the uied.templates file?

**Comment by taw** [comment]
The comments say the decoder is nowhere near finished, and that it's ETW only thing.

Looking at the code step one of unpacking `uied.templates` into individual files works (`templates_unpack` script), but stage two of parsing those individual files only extracts tga images from them and a few fields following that, and then gives up.

Did anyone else ever tried figuring out this format? It's been forever since I attempted that, so I don't really remember anything beyond what's in the code. I took a quick stab at many file formats, but usually gave up if it wasn't getting far.

**Comment by szmania** [comment]
I'm curious if the unpacking of uied.templates files unpacked ui files that can be converted to xml? 

In NTW we have in data/ui/templates/row_template_region for example, which is a ui file that can be converted to xml. I can't find its ETW counterpart. Therefore, I'm assuming its in ETW ui/templates/uied.templates. I believe this because ETW ui/templates/template.row_template_region.lua references xml objects (ui components) that I can't find in any of the unpacked ui > xml files within ETW. So I'm assuming they are in uied.templates.

**Comment by szmania** [comment]
Yeah you are right, these are .tga files, some of them, but they also have ui xml like data too. Can we parse those out from the tga files?

**Comment by taw** [comment]
So looking at those files, they look kinda like UI files, possibly the original version since they lack `Version0xx` header, and seem to have fewer fields.

Oldest version supported by UI file converter is `032`, so if that's de facto version `001`, that's a fairly awkward gap.

Given this theory, best way to approach it would be:

* creating a copy of `ui_converter/convert_ui.py` (that one's Python for fun)
* removing everything for versions > `032` and making sure it works for that - none of that will be relevant
* getting rid of `Version0xx` check, since it doesn't have version
* commenting out parts until it works
* (then maybe merging that back into ui converter)

It might be a bit tedious, but it could work just like that.

If all you need is just know if some ID from lua refers to some unpacked ui file, just check for strings inside it, those files use int32 numerical ids and ASCII strings, so it's a fairly simple check.

**Comment by szmania** [comment]
Interesting, I was kind of thinking of the same think, but was curious why there was no version tag within those ui files. 

All I needed was the ASCII strings (ui object names) for referencing ui objects in lua, so I ended up using the NTW xml file counterparts (which I'm assuming are nearly the same [for the UI component I'm specifically working on], textures differing between ETW and NTW). But if I do need to make UI edits in these fils I can't do in the lua UIComponent class, I will likely try your method.

Thanks.

**Comment by taw** [comment]
Is this issue still something we want to do anything with?

[There's now new UI converter than converts binary UI files to XML and back for all 10 Total War games](https://github.com/taw/etwng/tree/master/ui).

[There's much better online luac decompilers too](https://www.decompiler.com/).

Does that cover what you were after?
